cache/ringbuffer.go

```diff
@@ -21,46 +21,75 @@ import (
 )
 
 type ringBuffer struct {
-	buffer []*clientv3.Event
+	buffer []*batch
 	// head is the index immediately after the last non-empty entry in the buffer (i.e., the next write position).
 	head, tail, size int
 }
 
+// batch groups all events that share one ModRevision.
+type batch struct {
+	rev    int64
+	events []*clientv3.Event
+}
+
 type KeyPredicate = func([]byte) bool
 
 func newRingBuffer(capacity int) *ringBuffer {
 	// assume capacity > 0 – validated by Cache
 	return &ringBuffer{
-		buffer: make([]*clientv3.Event, capacity),
+		buffer: make([]*batch, capacity),
+	}
+}
+
+func (r *ringBuffer) Append(events []*clientv3.Event) {
+	start := 0
+	for end := 1; end < len(events); end++ {
+		if events[end].Kv.ModRevision != events[start].Kv.ModRevision {
+			r.append(events[start:end])
+			start = end
+		}
+	}
+	if start < len(events) {
+		r.append(events[start:])
```

Comment:
As `append` is an internal function we can use batch now.
```suggestion
func (r *ringBuffer) Append(events []*clientv3.Event) {
	start := 0
	for end := 1; end < len(events); end++ {
		if events[end].Kv.ModRevision != events[start].Kv.ModRevision {
			r.append(batch{rev: events[start].Kv.ModRevision, events: events[start:end]})
			start = end
		}
	}
	if start < len(events) {
		r.append(batch{rev: events[start].Kv.ModRevision, events: events[start:]})
```

Reply:
Thinking about copying objects is a premature optimization. Please note that passing argument via pointer still means you need to copy a pointer, which should have size of 8 bytes. Passing it via value is usually cheaper as most types are around the same size and there main cost is in dereferencing the pointer. For this case struct `batch` has int and a slice. Guessing this is around 4 x 64 bit int. Which might be 4 times more, but still pretty small. Please don't optimize code like this.

I usually decide whether to pass a pointer or value based on intention or convenience. Some rules out of my head.

If function is accepting a pointer it either:
* Wants to edit the value, and this be reflected in caller.
* Don't want to edit the value, but it is be more convenient as the type is always passed via pointer.

If function is accepting a value it:
* Wants to edit the value, but really doesn't want it reflected in caller.